capsules/extra/src/ieee802154/driver.rs

```diff
@@ -989,6 +1041,13 @@ impl<'a, M: device::MacDevice<'a>> SyscallDriver for RadioDriver<'a, M> {
                         |_| self.do_next_tx_sync(processid).into(),
                     )
             }
+            28 => self
+                .apps
+                .enter(processid, |app, _| {
+                    app.pending_rx = false;
```

Comment:
It seems like command 28 stops receive in a way. Why not just make this explicitly to declare "the rx buffer is new" so the kernel knows to reset the ring structure?

Reply:
When the kernel receives this command, the userprocess _should_ have already unallowed the buffer so the kernel has nothing it is able to even reset.

The motivation for having this was to prevent repeated upcalls. Previously, if multiple packets arrived quickly, two upcalls would be scheduled (the ring buffer avoids the upcalls being overwritten but does not guard against this). We do not want this to occur because this would mean the userprocess would be unable to transmit (since the receive function used both the receive and transmit upcall allocation). The `pending_rx` flag stops this from occuring. The side effect of this, however, is we need a mechanism to notify the kernel it can once again schedule an upcall (command 28).

@bradjc does this add some clarity for what this is doing? Would it be more clear to change this flag to `pending_rx_upcall`?

Reply:
Why not just unsubscribe the upcall?

Reply:
As I understand how unsubscribe works, I do not think this would solve the issue `pending_rx` and command 28 address (2 upcalls being scheduled while receiving). The current doubly scheduled receive upcall is happening when:

1. user process subscribes / allows buffer
2. kernel receives packet 1, places into buffer, and schedules upcall
3. kernel receives packet 2, places into buffer, and schedules another upcall (since the upcall count is two, two upcalls can be scheduled)

Unsubscribing the upcall would need to occur by the userprocess when the upcall is handled (at which time the second upcall has already been scheduled).

Reply:
The first argument to `schedule_upcall()` is the upcall index. So the driver having two upcalls means there are two separate upcalls, but each can be scheduled multiple times. The only limit is when the task queue for the process fills up (which right now has 10 spots https://github.com/tock/tock/blob/01918fac039fd6d37c4cea1b3bdbd9ac2683af3f/kernel/src/process_standard.rs#L1337).

I actually think unsubscribing _would_ solve the problem, as when the new subscribe (aka the unsubscribe) happens, we clear any pending upcalls for the old upcall (https://github.com/tock/tock/blob/01918fac039fd6d37c4cea1b3bdbd9ac2683af3f/kernel/src/kernel.rs#L977). Adding the command won't, since the upcall is still pending in the queue waiting for the app to call yield.

Reply:
I understand now. This is a great idea! I will make the changes to incorporate this instead of the command.